src/coordinator/api/v1/handler/prometheus/native/read.go

```diff
@@ -114,89 +104,94 @@ func (h *PromReadHandler) read(reqCtx context.Context, w http.ResponseWriter, pa
+			sortedBlockList, err = insertSortedBlock(blkResult.Block, sortedBlockList)
+			if err != nil {
+				return nil, err
+			}
 		}
 	}
 
-	return seriesMapToSeriesList(seriesMap)
+	seriesList, err := sortedBlocksToSeriesList(sortedBlockList)
+	for _, b := range sortedBlockList {
+		b.Close()
+	}
+
+	return seriesList, err
 }
 
-func insertSeriesInMap(blockSeries block.Series, seriesMap map[string][]block.Series) {
-	seriesID := blockSeries.Meta.Name
-	blockList, ok := seriesMap[seriesID]
-	if !ok {
-		seriesMap[seriesID] = make([]block.Series, 1)
-		seriesMap[seriesID][0] = blockSeries
-		return
+func sortedBlocksToSeriesList(blockList []block.Block) ([]*ts.Series, error) {
+	if len(blockList) == 0 {
+		return []*ts.Series{}, nil
 	}
 
-	// Insert sorted by start time
-	for idx, s := range blockList {
-		// TODO(nikunj): Use binary search once blocklist becomes big
-		if blockSeries.Meta.Bounds.Start.Before(s.Meta.Bounds.Start) {
-			blockList = append(blockList, block.Series{})
-			copy(blockList[idx+1:], blockList[idx:])
-			blockList[idx] = blockSeries
-			seriesMap[seriesID] = blockList
-			return
-		}
+	firstBlock := blockList[0]
+	numSeries := firstBlock.Series()
+	seriesList := make([]*ts.Series, numSeries)
+	seriesIters := make([]block.SeriesIter, len(blockList))
+	for i, b := range blockList {
+		seriesIters[i] = b.SeriesIter()
 	}
 
-	// If all start times lesser, then append to the end
-	seriesMap[seriesID] = append(blockList, blockSeries)
-}
+	seriesMeta := firstBlock.SeriesMeta()
+	bounds := firstBlock.Meta().Bounds
+	for i := 0; i < numSeries; i++ {
+		values := ts.NewFixedStepValues(bounds.StepSize, firstBlock.Steps()*len(blockList), math.NaN(), bounds.Start)
+		valIdx := 0
+		for idx, iter := range seriesIters {
+			if !iter.Next() {
+				return nil, fmt.Errorf("invalid number of datapoints for series: %d, block: %d", i, idx)
+			}
 
-func seriesMapToSeriesList(seriesMap map[string][]block.Series) ([]ts.Series, error) {
-	seriesList := make([]ts.Series, 0, len(seriesMap))
-	for _, blockSeriesList := range seriesMap {
-		s, err := blockSeriesListToSeries(blockSeriesList)
-		if err != nil {
-			return nil, err
+			blockSeries := iter.Current()
+			for i := 0; i < blockSeries.Len(); i++ {
+				values.SetValueAt(valIdx, blockSeries.ValueAtStep(i))
+				valIdx++
+			}
 		}
 
-		seriesList = append(seriesList, *s)
+		seriesList[i] = ts.NewSeries(seriesMeta[i].Name, values, seriesMeta[i].Tags)
+
 	}
 
 	return seriesList, nil
 }
 
-func blockSeriesListToSeries(series []block.Series) (*ts.Series, error) {
-	if len(series) == 0 {
-		return &ts.Series{}, nil
+func insertSortedBlock(b block.Block, blockList []block.Block) ([]block.Block, error) {
+	if len(blockList) == 0 {
+		blockList = append(blockList, b)
+		return blockList, nil
 	}
 
-	firstSeries := series[0]
-	var totalDatapoints int
-	for _, s := range series {
-		totalDatapoints += s.Len()
+	firstBlock := blockList[0]
+	if firstBlock.Series() != b.Series() {
```

Comment:
Can we rename `Series` to `SeriesCount` or something similar to give better context?